### PR TITLE
SONAR-10046 Add new constraints on ws api

### PR DESCRIPTION
**Pull request by gjambet**
SONAR-10046 Check minimum length of search parameters in WS

### CODE REVIEW

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [ListAction.java#L133](https://github.com/SonarSource/sonarqube/blob/57a55a8c548748584e2611d77f5c57b246459b5f/server/sonar-server/src/main/java/org/sonar/server/branch/ws/ListAction.java#L133): Specify a "BranchDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by julienlancelot** [comment]
This PR should not contains same commits as https://github.com/SonarSource/sonarqube/pull/2789, otherwise it's hard to review :(

**Comment by julienlancelot** [comment]
Commit message of https://github.com/SonarSource/sonarqube/pull/2797/commits/57a55a8c548748584e2611d77f5c57b246459b5f is wrong, the JIRA ticket should be SONAR-10046, and not SONAR-10040

File: server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentTreeAction.java [hunk 1]
```diff
@@ -309,8 +309,6 @@ private static ComponentTreeWsRequest toComponentTreeWsRequest(Request request)
       .setPageSize(request.mandatoryParamAsInt(Param.PAGE_SIZE))
       .setQuery(request.param(Param.TEXT_QUERY));
     String searchQuery = componentTreeWsRequest.getQuery();
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentTreeAction.java]: MAJOR Remove this useless assignment to local variable "searchQuery". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1854)
MINOR Remove this unused "searchQuery" local variable. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1481)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: server/sonar-server/src/main/java/org/sonar/server/permission/ws/GroupsAction.java [hunk 2]
```diff
@@ -81,11 +81,16 @@ public void define(WebService.NewController context) {
         "<li>'Administer' rights on the specified project</li>" +
         "</ul>")
       .addPagingParams(DEFAULT_PAGE_SIZE, RESULTS_MAX_SIZE)
-      .addSearchQuery("sonar", "names").setDescription("Limit search to group names that contain the supplied string. Must have at least %d characters.<br/>" +
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/permission/ws/GroupsAction.java]: You should not remove `addSearchQuery` but replace it by `createSearchQuery` in order to use `.setMinimumLength(SEARCH_QUERY_MIN_LENGTH)` on it

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/SuggestionsAction.java [hunk 3]
```diff
@@ -122,8 +122,8 @@ public void define(WebService.NewController context) {
 
     action.createParam(PARAM_QUERY)
       .setRequired(false)
-      .setDescription("Search query with a minimum of two characters. Can contain several search tokens, separated by spaces. " +
-        "Search tokens with only one character will be ignored.")
+      .setMinimumLength(2)
+      .setDescription("Can contain several search tokens, separated by spaces.")
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/SuggestionsAction.java]: `Search query` is missing in the description

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/TreeAction.java [hunk 4]
```diff
@@ -55,10 +55,10 @@
 import org.sonarqube.ws.client.component.TreeWsRequest;
 
 import static com.google.common.base.Preconditions.checkArgument;
-import static com.google.common.collect.FluentIterable.from;
 import static java.lang.String.CASE_INSENSITIVE_ORDER;
 import static java.lang.String.format;
 import static java.util.Collections.emptyMap;
+import static java.util.stream.Collectors.toList;
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/TreeAction.java]: In fact we're using our own Collectors (returning immutable lists) => please use MoreCollectors instead

**Comment by julienlancelot** [comment]
Please rebase to apply fixup https://github.com/SonarSource/sonarqube/pull/2797/commits/ec69df1e8b8d8936744702c090a98c4728806abe

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/TreeAction.java (hunk 4 above)
Comment (gjambet) [reply server/sonar-server/src/main/java/org/sonar/server/component/ws/TreeAction.java]: done

File: server/sonar-server/src/main/java/org/sonar/server/permission/ws/GroupsAction.java (hunk 2 above)
Comment (gjambet) [reply server/sonar-server/src/main/java/org/sonar/server/permission/ws/GroupsAction.java]: done

**Pull request merged**